### PR TITLE
Implement flatMapValues method

### PR DESCRIPTION
**Pull request by danosipov**
Sometimes its useful to have flatMapValues, similar to mapValues, but flatter.

### CODE REVIEW

File: scalding-core/src/main/scala/com/twitter/scalding/typed/KeyedList.scala [hunk 1]
```diff
@@ -174,6 +175,13 @@ trait KeyedListLike[K, +T, +This[K, +T] <: KeyedListLike[K, T, This]]
     mapGroup { (_, iter) => iter.map(fn) }
 
   /**
+   * Similar to mapValues, but works like flatMap, returning a collection of outputs
+   * for each value input.
+   */
+  def flatMapValues[V](fn: T => GenTraversableOnce[V]): This[K, V] =
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/KeyedList.scala]: we generally use `TraversableOnce[V]` here, can we stick to that style?

**Comment by johnynek** [comment]
can you add the same method to `TypedPipe`? we try to add all methods that make sense on both to both. Something like:

```scala
def flatMapValues[K, V1, V2](fn: V1 => TraversableOnce[V2])(implicit ev: T <:< (K, V1)): TypedPipe[(K, V2)] =
  raiseTo[K, V1].flatMap { case (k, v1) => fn(v1).map { v2 => k -> v2 } } 
```

**Comment by johnynek** [comment]
Thanks for adding this. You're right, it comes up, and summingbird added it a long time ago.

**Comment by danosipov** [comment]
@johnynek Added the method to TypedPipe, but there appear to be no tests for it. It type checks though :)

File: scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala [hunk 2]
```diff
@@ -300,6 +302,10 @@ trait TypedPipe[+T] extends Serializable {
   def mapValues[K, V, U](f: V => U)(implicit ev: T <:< (K, V)): TypedPipe[(K, U)] =
     raiseTo[(K, V)].map { case (k, v) => (k, f(v)) }
 
+  /** Similar to mapValues, but allows to return a collection of outputs for each input value */
+  def flatMapValues[K, V, U](f: V => TraversableOnce[U])(implicit ev: T <:< (K, V)): TypedPipe[(K, TraversableOnce[U])] =
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala]: I think we want `TypedPipe[(K, U)]` to be the return type. In other words, you have another copy of mapValues here, which you can use if you don't want the flattening, but I think this one should behave exactly as it does on KeyedListLike.

Comment (danosipov) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala]: But if I use `fn(v1).map { v2 => k -> v2 }` This returns a Traversable[U] (or Traversable[V2] from example above)

Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala]: look at my code. It had a flatMap yours does not have.

Comment (danosipov) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala]: :dizzy_face: Of course, fixed

**Comment by johnynek** [comment]
Thanks! This is useful.

**Pull request merged**